sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTypes.scala

```diff
@@ -43,16 +43,27 @@ private[parquet] object ParquetTypesConverter extends Logging {
   }
 
   /**
-   * Compute the FIXED_LEN_BYTE_ARRAY length needed to represent a given DECIMAL precision.
+   * BYTES_FOR_PRECISION computes the required bytes to store a value of a certain decimal
+   * precision.
    */
-  private[parquet] val BYTES_FOR_PRECISION = Array.tabulate[Int](38) { precision =>
-    var length = 1
+  private[parquet] def BYTES_FOR_PRECISION_COMPUTE(precision : Int) : Int = {
+    var length = (precision / math.log10(2) - 1).toInt / 8
     while (math.pow(2.0, 8 * length - 1) < math.pow(10.0, precision)) {
       length += 1
     }
     length
   }
 
+  private[parquet] def BYTES_FOR_PRECISION_STATIC =
+    (0 to 30).map(BYTES_FOR_PRECISION_COMPUTE).toArray
```

Comment:
30 should probably be replaced with 38, which fits in 16 bytes, and is the maximum precision supported in Hive.